File: app/src/main/kotlin/com/apex/trader/presentation/screen/scanner/ScannerScreen.kt
package com.apex.trader.presentation.screen.scanner

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Settings
import androidx.compose.material.icons.outlined.AccountBalance
import androidx.compose.material3.Button
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.apex.trader.domain.strategy.Signal
import com.apex.trader.domain.strategy.SignalSide
import com.apex.trader.presentation.component.ApexCard
import com.apex.trader.presentation.component.ConfidenceBadge
import com.apex.trader.presentation.component.SidePill
import com.apex.trader.presentation.theme.ApexBear
import com.apex.trader.presentation.theme.ApexBull
import com.apex.trader.presentation.theme.ApexTextMuted

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ScannerScreen(
    onSignalClick: (String) -> Unit,
    onPositionsClick: () -> Unit,
    onSettingsClick: () -> Unit,
    vm: ScannerViewModel = hiltViewModel()
) {
    val ui by vm.ui.collectAsStateWithLifecycle()

    LaunchedEffect(Unit) {
        if (ui.signals.isEmpty() && !ui.scanning) vm.scan()
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Apex Scanner") },
                actions = {
                    IconButton(onClick = onPositionsClick) {
                        Icon(Icons.Outlined.AccountBalance, "Positions")
                    }
                    IconButton(onClick = onSettingsClick) {
                        Icon(Icons.Default.Settings, "Settings")
                    }
                },
                colors = TopAppBarDefaults.topAppBarColors(
                    containerColor = MaterialTheme.colorScheme.background
                )
            )
        }
    ) { padding ->
        Column(
            modifier = Modifier.fillMaxSize().padding(padding)
        ) {
            ScanHeader(
                ui = ui,
                onScan = { vm.scan() },
                onCancel = { vm.cancel() }
            )

            if (ui.error != null) {
                Text(
                    text = "⚠ ${ui.error}",
                    color = MaterialTheme.colorScheme.error,
                    modifier = Modifier.padding(horizontal = 16.dp)
                )
            }

            if (ui.signals.isEmpty() && !ui.scanning) {
                EmptyState()
            } else {
                LazyColumn(
                    modifier = Modifier.fillMaxSize().padding(horizontal = 12.dp),
                    verticalArrangement = Arrangement.spacedBy(8.dp),
                    contentPadding = androidx.compose.foundation.layout.PaddingValues(vertical = 8.dp)
                ) {
                    items(ui.signals, key = { it.symbol }) { sig ->
                        SignalCard(sig, onClick = { onSignalClick(sig.symbol) })
                    }
                }
            }
        }
    }
}

@Composable
private fun ScanHeader(
    ui: ScannerUiState,
    onScan: () -> Unit,
    onCancel: () -> Unit
) {
    Column(modifier = Modifier.padding(16.dp)) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = if (ui.scanning) "Scanning…" else "${ui.signals.size} signals",
                    style = MaterialTheme.typography.titleLarge
                )
                Text(
                    text = if (ui.scanning) "${ui.processed}/${ui.total} symbols • ${ui.current ?: ""}"
                    else if (ui.lastScanAt != null) "Last scan ${formatAgo(ui.lastScanAt)}"
                    else "Tap scan to begin",
                    color = ApexTextMuted,
                    style = MaterialTheme.typography.bodyMedium
                )
            }
            if (ui.scanning) {
                OutlinedButton(onClick = onCancel) { Text("Cancel") }
            } else {
                Button(onClick = onScan) { Text("Scan now") }
            }
        }
        if (ui.scanning && ui.total > 0) {
            Spacer(Modifier.height(8.dp))
            LinearProgressIndicator(
                progress = { ui.processed / ui.total.toFloat() },
                modifier = Modifier.fillMaxWidth().height(4.dp)
            )
        }
    }
}

@Composable
private fun SignalCard(sig: Signal, onClick: () -> Unit) {
    ApexCard(modifier = Modifier.clickable { onClick() }) {
        Column {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(sig.symbol, style = MaterialTheme.typography.titleLarge)
                Spacer(Modifier.padding(end = 10.dp))
                SidePill(if (sig.side == SignalSide.LONG) "LONG" else "SHORT")
                Spacer(Modifier.weight(1f))
                ConfidenceBadge(sig.confidence)
            }
            Spacer(Modifier.height(10.dp))
            Row {
                MetricSmall("Entry", "%.4f".format(sig.plan.entry))
                Spacer(Modifier.padding(end = 14.dp))
                MetricSmall(
                    "SL",
                    "%.4f".format(sig.plan.stopLoss),
                    color = ApexBear
                )
                Spacer(Modifier.padding(end = 14.dp))
                MetricSmall(
                    "TP1 / TP3",
                    "%.4f / %.4f".format(sig.plan.takeProfit1, sig.plan.takeProfit3),
                    color = ApexBull
                )
            }
            Spacer(Modifier.height(8.dp))
            Row {
                MetricSmall("ADX", "%.1f".format(sig.adx))
                Spacer(Modifier.padding(end = 14.dp))
                MetricSmall("RSI", "%.1f".format(sig.rsi))
                Spacer(Modifier.padding(end = 14.dp))
                MetricSmall("Vol×", "%.2f".format(sig.volumeSurge))
                Spacer(Modifier.padding(end = 14.dp))
                MetricSmall("Reasons", "${sig.reasonsPassed.size}/${sig.reasons.size}")
            }
        }
    }
}

@Composable
private fun MetricSmall(label: String, value: String, color: androidx.compose.ui.graphics.Color = MaterialTheme.colorScheme.onSurface) {
    Column {
        Text(label, color = ApexTextMuted, style = MaterialTheme.typography.labelSmall)
        Text(value, color = color, style = MaterialTheme.typography.labelMedium)
    }
}

@Composable
private fun EmptyState() {
    Box(modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            Text("No high-confidence setups", style = MaterialTheme.typography.titleMedium)
            Text(
                "The market may be ranging — try again later or lower the confidence threshold in Settings.",
                color = ApexTextMuted,
                style = MaterialTheme.typography.bodyMedium,
                modifier = Modifier.padding(horizontal = 24.dp)
            )
        }
    }
}

private fun formatAgo(ms: Long): String {
    val sec = (System.currentTimeMillis() - ms) / 1000
    return when {
        sec < 60 -> "${sec}s ago"
        sec < 3600 -> "${sec / 60}m ago"
        else -> "${sec / 3600}h ago"
    }
}

File: lib/features/scanner/scanner_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../core/theme.dart';
import '../../domain/strategy.dart';
import '../../services/foreground_service.dart';
import '../../services/notification_service.dart';
import '../../widgets/common.dart';
import 'scanner_controller.dart';

class ScannerScreen extends ConsumerStatefulWidget {
  const ScannerScreen({
    super.key,
    required this.onSignalTap,
    required this.onPositionsTap,
    required this.onSettingsTap,
    required this.onJournalTap,
    required this.onBacktestTap,
  });

  final void Function(String symbol) onSignalTap;
  final VoidCallback onPositionsTap;
  final VoidCallback onSettingsTap;
  final VoidCallback onJournalTap;
  final VoidCallback onBacktestTap;

  @override
  ConsumerState<ScannerScreen> createState() => _ScannerScreenState();
}

class _ScannerScreenState extends ConsumerState<ScannerScreen> {
  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addPostFrameCallback((_) {
      final s = ref.read(scannerControllerProvider);
      if (s.signals.isEmpty && !s.scanning) {
        ref.read(scannerControllerProvider.notifier).scan();
      }
    });
  }

  Future<void> _minimize(BuildContext context) async {
    final granted = await NotificationService.instance.areNotificationsAllowed();
    if (!granted) {
      final ok = await NotificationService.instance.requestNotificationPermission();
      if (!ok && context.mounted) {
        ScaffoldMessenger.of(context).showSnackBar(
          const SnackBar(
            content: Text(
              'Notification permission needed to keep the app running in the background.',
            ),
          ),
        );
        return;
      }
    }
    final started = await ApexForegroundService.instance.start();
    if (!started && context.mounted) {
      ScaffoldMessenger.of(context).showSnackBar(
        const SnackBar(content: Text("Couldn't start the persistent notification.")),
      );
      return;
    }
    await ApexForegroundService.instance.minimize();
  }

  @override
  Widget build(BuildContext context) {
    final s = ref.watch(scannerControllerProvider);
    final ctrl = ref.read(scannerControllerProvider.notifier);

    return Scaffold(
      appBar: AppBar(
        title: const Text('Apex Scanner'),
        actions: [
          IconButton(
            tooltip: 'Minimize (keep running)',
            onPressed: () => _minimize(context),
            icon: const Icon(Icons.minimize),
          ),
          IconButton(
            tooltip: 'Backtest',
            onPressed: widget.onBacktestTap,
            icon: const Icon(Icons.history_toggle_off_outlined),
          ),
          IconButton(
            tooltip: 'Journal',
            onPressed: widget.onJournalTap,
            icon: const Icon(Icons.menu_book_outlined),
          ),
          IconButton(
            tooltip: 'Positions',
            onPressed: widget.onPositionsTap,
            icon: const Icon(Icons.account_balance_wallet_outlined),
          ),
          IconButton(
            tooltip: 'Settings',
            onPressed: widget.onSettingsTap,
            icon: const Icon(Icons.settings_outlined),
          ),
        ],
      ),
      body: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          Padding(
            padding: const EdgeInsets.fromLTRB(16, 12, 16, 8),
            child: Row(
              children: [
                Expanded(
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(
                        s.scanning ? 'Scanning…' : '${s.signals.length} signals',
                        style: Theme.of(context).textTheme.titleLarge,
                      ),
                      Text(
                        s.scanning
                            ? '${s.processed}/${s.total} symbols • ${s.current ?? ''}'
                            : s.lastScanAt != null
                                ? 'Last scan ${_relTime(s.lastScanAt!)}'
                                : 'Tap scan to begin',
                        style: const TextStyle(color: ApexColors.textMuted),
                      ),
                    ],
                  ),
                ),
                if (s.scanning)
                  OutlinedButton(onPressed: () {}, child: const Text('Cancel'))
                else
                  ElevatedButton(onPressed: ctrl.scan, child: const Text('Scan now')),
              ],
            ),
          ),
          if (s.scanning && s.total > 0)
            Padding(
              padding: const EdgeInsets.symmetric(horizontal: 16),
              child: LinearProgressIndicator(
                value: s.processed / s.total,
                color: ApexColors.primary,
                backgroundColor: ApexColors.surfaceVariant,
                minHeight: 4,
              ),
            ),
          if (s.error != null)
            Padding(
              padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 8),
              child: Text('⚠ ${s.error}', style: const TextStyle(color: ApexColors.bear)),
            ),
          if (s.autoTradePlaced.isNotEmpty || s.autoTradeWarnings.isNotEmpty)
            Padding(
              padding: const EdgeInsets.fromLTRB(16, 4, 16, 4),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  ...s.autoTradePlaced.map(
                    (p) => Text('✓ Auto-traded $p',
                        style: const TextStyle(color: ApexColors.bull, fontSize: 12.5)),
                  ),
                  ...s.autoTradeWarnings.map(
                    (w) => Text('⚠ $w',
                        style: const TextStyle(color: ApexColors.highlight, fontSize: 12.5)),
                  ),
                ],
              ),
            ),
          Expanded(
            child: s.signals.isEmpty && !s.scanning
                ? _empty(context)
                : ListView.separated(
                    padding: const EdgeInsets.fromLTRB(12, 4, 12, 12),
                    itemCount: s.signals.length,
                    separatorBuilder: (_, __) => const SizedBox(height: 8),
                    itemBuilder: (_, i) {
                      final sig = s.signals[i];
                      return _SignalCard(
                        signal: sig,
                        onTap: () => widget.onSignalTap(sig.symbol),
                      );
                    },
                  ),
          ),
        ],
      ),
    );
  }

  Widget _empty(BuildContext context) {
    return Center(
      child: Padding(
        padding: const EdgeInsets.all(24),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Text('No high-confidence setups', style: Theme.of(context).textTheme.titleMedium),
            const SizedBox(height: 6),
            const Text(
              'The market may be ranging — try again later or lower the confidence threshold in Settings.',
              textAlign: TextAlign.center,
              style: TextStyle(color: ApexColors.textMuted),
            ),
          ],
        ),
      ),
    );
  }
}

class _SignalCard extends StatelessWidget {
  const _SignalCard({required this.signal, required this.onTap});
  final Signal signal;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    final p = signal.plan;
    return ApexCard(
      onTap: onTap,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            children: [
              Text(signal.symbol,
                  style: const TextStyle(fontSize: 18, fontWeight: FontWeight.w700)),
              const SizedBox(width: 10),
              SidePill(side: signal.side == SignalSide.long ? 'LONG' : 'SHORT'),
              const Spacer(),
              ConfidenceBadge(score: signal.confidence),
            ],
          ),
          const SizedBox(height: 10),
          Wrap(
            runSpacing: 6,
            spacing: 18,
            children: [
              _Metric(label: 'Entry', value: p.entry.toStringAsFixed(4)),
              _Metric(label: 'SL', value: p.stopLoss.toStringAsFixed(4), color: ApexColors.bear),
              _Metric(
                label: 'TP1 / TP3',
                value: '${p.takeProfit1.toStringAsFixed(4)} / ${p.takeProfit3.toStringAsFixed(4)}',
                color: ApexColors.bull,
              ),
              _Metric(label: 'ADX', value: signal.adx.toStringAsFixed(1)),
              _Metric(label: 'RSI', value: signal.rsi.toStringAsFixed(1)),
              _Metric(label: 'Vol×', value: signal.volumeSurge.toStringAsFixed(2)),
              _Metric(
                label: 'Reasons',
                value:
                    '${signal.reasonsPassed.length}/${signal.reasons.length}',
              ),
            ],
          ),
        ],
      ),
    );
  }
}

class _Metric extends StatelessWidget {
  const _Metric({required this.label, required this.value, this.color});
  final String label;
  final String value;
  final Color? color;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      mainAxisSize: MainAxisSize.min,
      children: [
        Text(label,
            style: const TextStyle(color: ApexColors.textMuted, fontSize: 11)),
        Text(value,
            style: TextStyle(
              color: color ?? ApexColors.text,
              fontFamily: 'monospace',
              fontSize: 12,
              fontWeight: FontWeight.w500,
            )),
      ],
    );
  }
}

String _relTime(int ms) {
  final sec = (DateTime.now().millisecondsSinceEpoch - ms) ~/ 1000;
  if (sec < 60) return '${sec}s ago';
  if (sec < 3600) return '${sec ~/ 60}m ago';
  return '${sec ~/ 3600}h ago';
}
